pkg/k8s/endpointsynchronizer/cep.go

```diff
@@ -180,77 +185,108 @@ func (epSync *EndpointSynchronizer) RunK8sCiliumEndpointSync(e *endpoint.Endpoin
-					mdl.DeepCopyInto(&cep.Status)
-					var err2 error
+				// We have no localCEP copy. We need to fetch it for updates, below.
+				// This is unexpected as there should be only 1 writer per CEP, this
+				// controller, and the localCEP created on startup will be used.
+				if localCEP == nil {
+					localCEP, err = ciliumClient.CiliumEndpoints(namespace).Get(podName, meta_v1.GetOptions{})
 					switch {
-					case ciliumUpdateStatusVerConstr.Check(k8sServerVer):
-						_, err2 = ciliumClient.CiliumEndpoints(namespace).UpdateStatus(cep)
-					default:
-						_, err2 = ciliumClient.CiliumEndpoints(namespace).Update(cep)
-					}
-					if err2 != nil {
-						scopedLog.WithError(err2).Error("Cannot update CEP")
-						return err2
-					}
+					// The CEP doesn't exist in k8s. This is unexpetected but may occur
+					// if the endpoint was removed from k8s but not yet within the agent.
+					// Mark the CEP for creation on the next controller iteration. This
+					// may never occur if the controller is stopped on Endpoint delete.
+					case err != nil && k8serrors.IsNotFound(err):
+						needInit = true
+						return err
 
-					lastMdl = mdl
-					return nil
+					// We cannot read the upstream CEP. needInit will cause the next
+					// iteration to delete and create the CEP. This is an unexpected
+					// situation.
+					case err != nil && k8serrors.IsInvalid(err):
+						scopedLog.WithError(err).Warn("Invalid CEP during update")
+						needInit = true
+						return nil
+
+					// A real error
+					case err != nil:
+						scopedLog.WithError(err).Error("Cannot get CEP during update")
+						return err
+					}
 				}
 
-				// The CEP was not found, this is the first creation of the endpoint
-				cep = &cilium_v2.CiliumEndpoint{
-					ObjectMeta: meta_v1.ObjectMeta{
-						Name: podName,
-					},
-					Status: *mdl,
+				// We have an object to reuse. Update and push it up. In the case of an
+				// update error, we retry in the next iteration of the controller using
+				// the copy returned by Update.
+				scopedLog.Debug("Updating CEP from local copy")
+				// Update the copy of the cep
+				mdl.DeepCopyInto(&localCEP.Status)
+				switch {
+				case ciliumUpdateStatusVerConstr.Check(k8sServerVer):
+					localCEP, err = ciliumClient.CiliumEndpoints(namespace).UpdateStatus(localCEP)
+				default:
+					localCEP, err = ciliumClient.CiliumEndpoints(namespace).Update(localCEP)
 				}
 
-				_, err = ciliumClient.CiliumEndpoints(namespace).Create(cep)
-				if err != nil {
-					scopedLog.WithError(err).Error("Cannot create CEP")
+				// Handle Update errors or return successfully
+				switch {
+				// Return no error when we see a conflict. We want to retry without a
+				// backoff and the Update* calls returned the current localCEP
+				case err != nil && k8serrors.IsConflict(err):
+					scopedLog.WithError(err).Info("Cannot update CEP due to a revision conflict. The next controller execution will try again")
```

Comment:
I still struggle with this info message. Is it an expected condition, if so, debug. Is it unexpected and rare -> warning.

Reply:
It is very unexpected. It would mean something else is updating the CEP and that should never happen (even the GC would only happen if the endpoint is deleted so this controller would be stopped anyway). I'll change it to a warning.